### PR TITLE
Add ScopedEntityMongoUtils

### PR DESCRIPTION
**Pull request by kingzacko1**
Since we are moving away from using `PaginatedDbService` we should do the same for `ScopedDbService` which extends the `PaginatedDbService` and adds some mutability checks. 

/nocl not user facing
## Description

Creates a `ScopedEntityMongoUtils` with `ScopedDbService`'s mutability checks to use in new DB service classes for `ScopedEntity` objects.
 
## Motivation and Context

moving towards using new Mongo API instead of deprecated mongojack library.

## How Has This Been Tested?

local dev env
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog2/database/utils/ScopedEntityMongoUtils.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        this.entityScopeService = entityScopeService;
+    }
+
+    /**
+     * Performs a valid scope and mutability check before updating an existing entity.
+     *
+     * @param entity ScopedEntity to be updated
+     * @return the newly updated entity
+     */
+    public T update(T entity) {
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/database/utils/ScopedEntityMongoUtils.java:44]: Please add unit tests for this class. Thanks!

File: graylog2-server/src/main/java/org/graylog2/database/MongoCollections.java [hunk 2]
```diff
@@ -93,4 +109,11 @@ public <T extends MongoEntity> MongoUtils<T> utils(String collectionName, Class<
     public <T extends MongoEntity> MongoUtils<T> utils(MongoCollection<T> collection) {
         return new MongoUtils<>(collection, objectMapper);
     }
+
+    /**
+     * Provides typical utility methods and mutability checks for ScopedEntity objects
+     */
+    public <T extends ScopedEntity> ScopedEntityMongoUtils<T> scopedEntityUtils(MongoCollection<T> collection) {
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/database/MongoCollections.java:116]: What do you think about adding the `EntityScopeService` as a second parameter to this method? That way, we don't have to inject the service into `MongoCollections` and can keep its dependencies lean.

File: graylog2-server/src/main/java/org/graylog2/database/entities/ScopedEntity.java [hunk 3]
```diff
@@ -26,11 +27,11 @@
  * Entity base class, which can be used to enforce that each entity implementation
  * has the required id and _scope fields.
  */
-
-public abstract class ScopedEntity {
+public abstract class ScopedEntity implements MongoEntity {
     public static final String FIELD_ID = "id";
     public static final String FIELD_SCOPE = "_scope";
 
+    @Override
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/database/entities/ScopedEntity.java:34]: You can remove the `id()` method from this abstract class. Auto-value will pull it in from the `MongoEntity` interface.

**Review by bernd — CHANGES_REQUESTED** [review]

File: graylog2-server/src/main/java/org/graylog2/database/utils/MongoUtils.java [hunk 4]
```diff
@@ -99,6 +101,26 @@ public static Bson idEq(@Nonnull ObjectId id) {
         return Filters.eq("_id", id);
     }
 
+    /**
+     * Create a query constraint to match a document's ID against the given list of Hex strings.
+     *
+     * @param ids Collection of hex string representations of an {@link ObjectId}
+     * @return An 'in' filter.
+     */
+    public static Bson stringIdsIn(Collection<String> ids) {
```
[the remark this answers is not in the document]
Comment (kingzacko1) [reply graylog2-server/src/main/java/org/graylog2/database/utils/MongoUtils.java:110]: @bernd I did try this naming convention originally, but received this compile time error due to generics. I googled some workaround to keep the name convention but nothing seemed to fit for this particular case/. 
image

[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/main/java/org/graylog2/database/utils/MongoUtils.java:110]: Ah okay. Sorry, I didn't think about that.

File: graylog2-server/src/main/java/org/graylog2/database/MongoCollections.java (hunk 2 above)
Comment (kingzacko1) [reply graylog2-server/src/main/java/org/graylog2/database/MongoCollections.java:116]: Definitely like that better. Thanks for the suggestion.

**Review by danotorrey — APPROVED** [review]
LGTM and tests successfully.

**Review by ryan-carroll-graylog — APPROVED** [review]
Looks good (aside from Bernd's last comment) and tests successfully! Nice job tackling this!

File: graylog2-server/src/main/java/org/graylog2/database/entities/ScopedEntity.java [hunk 5]
```diff
@@ -17,26 +17,18 @@
...
-import javax.annotation.Nullable;
-
 /**
  * Entity base class, which can be used to enforce that each entity implementation
  * has the required id and _scope fields.
  */
-
-public abstract class ScopedEntity {
+public abstract class ScopedEntity implements MongoEntity {
     public static final String FIELD_ID = "id";
```
[the remark this answers is not in the document]
Comment (kingzacko1) [reply graylog2-server/src/main/java/org/graylog2/database/entities/ScopedEntity.java:29]: It's used in the AbstractBuilder below and referenced in a few other places either by directly importing it or referencing a subclass. 
image

**Review by bernd — APPROVED** [review]

**Pull request merged**